src/arch/cuda/hybridAPI/hapi_impl.cpp

```diff
@@ -89,15 +93,89 @@ CsvDeclare(GPUManager, gpu_manager);
 CpvDeclare(int, my_device); // GPU device that this thread is mapped to
 CpvDeclare(bool, device_rep); // Is this PE a device representative thread? (1 per device)
 
+// Returns the local rank of the logical node (process) that the given PE belongs to
+static inline int CmiNodeRankLocal(int pe) {
+  // Logical node index % Number of logical nodes per physical node
+  return CmiNodeOf(pe) % (CmiNumNodes() / CmiNumPhysicalNodes());
+}
+
+// Returns the local rank of the logical node that I belong to
+static inline int CmiMyNodeRankLocal() {
+  return CmiNodeRankLocal(CmiMyPe());
+}
+
+// HAPI internal function declarations
+static void hapiInitCsv();
+static void hapiInitCpv();
+static void hapiExitCsv();
+
+static void hapiMapping(char** argv);
+static void hapiRegisterCallbacks();
+
+// CUDA IPC related functions
+static void shmCreate();
+static void shmCleanup();
+static void ipcHandleCreate();
+static void ipcHandleOpen();
+
+// Called by all PEs in Charm++ layer init
+void hapiInit(char** argv) {
+  if (!CmiInCommThread()) {
+    if (CmiMyRank() == 0) {
+      hapiInitCsv(); // Initialize per-process variables (GPUManager)
+    }
+    hapiInitCpv(); // Initialize per-PE variables
+
+    CmiNodeBarrier();
```

Comment:
Can you add comments as to why a NodeBarrier is needed after a set of operations? (Pardon my lack of within-node/device level stuff).

Reply:
Will do